sdks/python/apache_beam/io/watermark_estimators.py

```diff
@@ -0,0 +1,101 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""A collection of WatermarkEstimator implementations that SplittableDoFns
+can use."""
+
+# pytype: skip-file
+
+from __future__ import absolute_import
+
+from apache_beam.io.iobase import WatermarkEstimator
+from apache_beam.utils.timestamp import Timestamp
+
+
+class MonotonicWatermarkEstimator(WatermarkEstimator):
+  """A WatermarkEstimator which assumes that timestamps of all ouput records
+  are increasing monotonically.
+  """
+  def __init__(self, timestamp):
+    """For a new <element, restriction> pair, the initial value is None. When
+    resuming processing, the initial timestamp will be the last reported
+    watermark.
+    """
+    self._watermark = timestamp
+
+  def observe_timestamp(self, timestamp):
+    if self._watermark is None:
+      self._watermark = timestamp
+    else:
+      if timestamp < self._watermark:
+        raise ValueError('A MonotonicWatermarkEstimator expects output '
```

Comment:
Here's a conundrum: it might not be very easy for the user to actually catch and act on this error. (For example, say they're reading log entries and they think it's ordered but one occasionally goes backward). There are three options: 

(1) Stop the pipeline.
(2) Suppress the error and emit the item as possibly late data.
(3) Move the timestamp forward to respect the watermark. 

Should this be configurable? Note that option (3) would require a return value here.

Reply:
This is an interesting question. 
I prefer stopping pipeline by raising error here. I think observing timestamp and emitting late data should be done inside the DoFn boday by the SDF author, given that `MonotonicWatermarkEstimator.observe_timestamp()` is called by the system. We can add more docstring to make option1 more explicit.

Reply:
The problem is that the user *can't* do (2) or (3) themselves. Let's at least add a JIRA to consider making this configurable (either here or as alternative estimators) but not block on this. (1) is certainly the safest default.